File: problems/树/145.二叉树的后序遍历.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <utility>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

vector<int> postorderTraversal(TreeNode* root) {
    vector<int> result;
    stack<pair<TreeNode*, bool>> st;
    if (root) st.push({root, false});
    while (!st.empty()) {
        auto [node, visited] = st.top(); st.pop();
        if (visited) {
            result.push_back(node->val);
        } else {
            st.push({node, true});
            if (node->right) st.push({node->right, false});
            if (node->left)  st.push({node->left, false});
        }
    }
    return result;
}
int main() {
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(3);
    for (int v : postorderTraversal(root)) cout << v << " "; cout << endl;  // 3 2 1
    for (int v : postorderTraversal(new TreeNode(1))) cout << v << " "; cout << endl;  // 1
    for (int v : postorderTraversal(nullptr)) cout << v << " "; cout << endl;  // (empty)
    return 0;
}

File: problems/树/145.二叉树的后序遍历.py
#!/usr/bin/env python3
"""
145. 二叉树的后序遍历
"""
from typing import List, Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def postorder_traversal(root: Optional[TreeNode]) -> List[int]:
    result = []
    stack = []
    cur = root
    while cur or stack:
        while cur:
            stack.append((cur, False))
            cur = cur.left
        node, visited = stack.pop()
        if visited:
            result.append(node.val)
        else:
            stack.append((node, True))
            cur = node.right
    return result

if __name__ == "__main__":
    root1 = TreeNode(1, None, TreeNode(2, TreeNode(3), None))
    print(postorder_traversal(root1))  # [3, 2, 1]
    print(postorder_traversal(TreeNode(1)))  # [1]
    print(postorder_traversal(None))  # []
